packages/firestore/src/model/transform_operation.ts

```diff
@@ -141,6 +158,62 @@ export class ArrayRemoveTransformOperation implements TransformOperation {
   }
 }
 
+/**
+ * Implements the backend semantics for locally computed NUMERIC_ADD transforms.
+ * Converts all field values to integers or doubles, but unlike the backend does
+ * not cap integer values at 2^63. Instead, JavaScript number arithmetic is used
+ * and precision loss can occur for values greater than 2^53.
+ */
+export class NumericAddTransformOperation implements TransformOperation {
+  readonly isIdempotent = false;
+
+  constructor(readonly operand: NumberValue) {}
+
+  applyToLocalView(
+    previousValue: FieldValue | null,
+    localWriteTime: Timestamp
+  ): FieldValue {
+    // PORTING NOTE: Since JavaScript's integer arithmetic is limited to 53 bit
+    // precision and resolves overflows by reducing precision, we do not manually
+    // cap overflows at 2^63.
+
+    // Return an integer value iff the previous value and the operand is an
+    // integer.
+    if (
+      previousValue instanceof IntegerValue &&
+      this.operand instanceof IntegerValue
+    ) {
+      const sum = previousValue.internalValue + this.operand.internalValue;
+      return new IntegerValue(sum);
+    } else if (previousValue instanceof NumberValue) {
+      const sum = previousValue.internalValue + this.operand.internalValue;
+      return new DoubleValue(sum);
+    }
+
+    // If the existing value is not a number, use the value of the transform as
```

Comment:
Can you put this in an else block?  (since I don't like early return and conceptually I think there are 3 possible cases here and we should just structure the code as such)

Reply:
Done